pkg/istio-agent/sds-agent.go

```diff
@@ -0,0 +1,405 @@
+	// If set SDS will be used - either local or via hostPath.
+	JWTPath string
+
+	// SDSAddress is the address of the SDS server. Starts with unix: for hostpath mount or built-in
+	// May also be a https address.
+	SDSAddress string
+
+	// CertPath is set with the location of the certs, or empty if mounted certs are not present.
+	CertsPath string
+
+	// RequireCerts is set if the agent requires certificates:
+	// - if controlPlaneAuthEnabled is set
+	// - port of discovery server is not 15010 (the plain text default).
+	RequireCerts bool
+}
+
+// DetectSDS will attempt to find nodeagent SDS and token. If not found will attempt to find
+// a location to start SDS.
+//
+// If node agent and JWT are mounted: it indicates user injected a config using hostPath, and will be used.
+//
+func DetectSDS(discAddr string, tlsRequired bool) *AgentConf {
+	ac := &AgentConf{}
+	if _, err := os.Stat(JWTPath); err == nil {
+		ac.JWTPath = JWTPath
+	}
+
+	hostSDSUDS := sdsUdsPathVar.Get()
+	if strings.HasPrefix(hostSDSUDS, "unix:") {
+		// Skip the prefix
+		hostSDSUDS = sdsUdsPathVar.Get()[5:]
+		if _, err := os.Stat(hostSDSUDS); err == nil {
+			ac.SDSAddress = hostSDSUDS
+		}
+	}
+
+	if _, err := os.Stat("/etc/certs/key.pem"); err == nil {
+		ac.CertsPath = "/etc/certs"
+	}
+	if tlsRequired {
+		ac.RequireCerts = true
+	}
+
+	return ac
+}
+
+// Simplified SDS setup. This is called if and only if user has explicitly mounted a K8S JWT token, and is not
+// using a hostPath mounted or external SDS server.
+//
+// 1. External CA: requires authenticating the trusted JWT AND validating the SAN against the JWT.
+//    For example Google CA
+//
+// 2. Indirect, using istiod: using K8S cert.
+//
+// 3. Monitor mode - watching secret in same namespace ( Ingress)
+//
+// 4. TODO: File watching, for backward compat/migration from mounted secrets.
+func StartSDS(conf *AgentConf, isSidecar bool) (*sds.Server, error) {
+	applyEnvVars()
+
+	gatewaySdsCacheOptions = workloadSdsCacheOptions
+
+	// Next to the envoy config, writeable dir (mounted as mem)
+	serverOptions.WorkloadUDSPath = LocalSDS
+
+	// TODO: remove the caching, workload has a single cert
+	workloadSecretCache, _ := newSecretCache(serverOptions)
+
+	var gatewaySecretCache *cache.SecretCache
+	if !isSidecar {
+		serverOptions.EnableIngressGatewaySDS = true
+		// TODO: what is the setting for ingress ?
+		serverOptions.IngressGatewayUDSPath = serverOptions.WorkloadUDSPath + "_ROUTER"
+		gatewaySecretCache = newIngressSecretCache(serverOptions)
+	}
+
+	// For sidecar and ingress we need to first get the certificates for the workload.
+	// We'll also save them in files, for backward compat with servers generating files
+	// TODO: use caClient.CSRSign() directly
+
+	// fail hard if we need certs ( control plane security enabled ) and we don't have mounted certs and
+	// we fail to load SDS
+	fail := conf.RequireCerts && conf.CertsPath == ""
+
+	tok, err := ioutil.ReadFile(conf.JWTPath)
+	if err != nil && fail {
```

Comment:
`err` is not handled here if `fail == false` which will probably break L233

Reply:
Done